src/lib/packager/product_patterns.rb

```diff
@@ -0,0 +1,146 @@
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# ------------------------------------------------------------------------------
+#
+
+require "yast"
+
+module Yast
+
+  # Evaluate the default patterns for the currently selected products
+  class ProductPatterns
+    include Yast::Logger
+
+    def initialize
+      Yast.import "Pkg"
+    end
+
+    # Find the default patterns for all selected products.
+    # @note The package management and the products needs to be initialized and
+    #   selected *before* using this method.
+    # @return [Array<String>] pattern names
+    def names
+      @names ||= find
+    end
+
+    # Select the default patterns to installation
+    # @note The package management and the products needs to be initialized and
+    #   selected *before* using this method.
+    # @return [Boolean] true if all patterns were successfuly selected,
+    #   false otherwise
+    def select
+      # use map + all? to try installing *all* patterns, plain all? would stop
+      # at the first failure
+      names.map { |p| Yast::Pkg.ResolvableInstall(p, :pattern) }.all?
+    end
+
+  private
+
+    # Find the default patterns for all selected products.
+    # @return [Array<String>] pattern names
+    def find
+      products = Yast::Pkg.ResolvableProperties("", :product, "")
+      remove_unselected(products)
+      products.map! { |product| product["name"] }
+      log.info "Found selected products: #{products}"
+
+      patterns = products.map { |p| product_patterns(p) }.flatten.uniq
+      log.info "Default patterns for the selected products: #{patterns.inspect}"
+
+      patterns
+    end
+
+    # Find the default patterns for the product.
+    # @param [String] product product name
+    # @return [Array<String>] pattern names
+    def product_patterns(product)
+      product_dependencies = dependencies(product)
+      product_provides = provides(product_dependencies)
+
+      default_patterns(product_provides)
+    end
+
+    # Find dependencies for the product (it's product package).
+    # @param [String] product product name
+    # @return [Array<Hash>] product dependencies, e.g. [{"provides" => "foo"},
+    #   {"requires" => "bar"}, ...]
+    def dependencies(product)
+      product_dependencies = []
+
+      resolvables = Yast::Pkg.ResolvableProperties(product, :product, "")
+      remove_unselected(resolvables)
+
+      resolvables.each do |resolvable|
+        prod_pkg = resolvable["product_package"]
+        next unless prod_pkg
+
+        release_resolvables = Yast::Pkg.ResolvableDependencies(prod_pkg, :package, "")
+        remove_unselected(release_resolvables)
+
+        release_resolvables.each do |release_resolvable|
+          deps = release_resolvable["deps"]
+          product_dependencies.concat(deps) if deps
+        end
+      end
+
+      log.debug "Product #{product} depependencies: #{product_dependencies}"
+
+      product_dependencies
+    end
```

Comment:
I have reformatted the files with Rubocop, it found several issues, but it did not complain at this place. Let's consider is as OK :wink: